hail/python/hailtop/aiocloud/aiogoogle/client/storage_client.py

```diff
@@ -55,32 +55,41 @@ async def __anext__(self):
 
 
 class InsertObjectStream(WritableStream):
-    def __init__(self, it, request_task):
+    def __init__(self,
+                 it: FeedableAsyncIterable[bytes],
+                 request_task: asyncio.Future):  # in Python 3.9: asyncio.Future[aiohttp.ClientResponse]
         super().__init__()
         self._it = it
         self._request_task = request_task
         self._value = None
 
     async def write(self, b):
         assert not self.closed
-        await self._it.feed(b)
-        return len(b)
+
+        fut = asyncio.ensure_future(self._it.feed(b))
+        try:
+            await asyncio.wait([fut, self._request_task], return_when=asyncio.FIRST_COMPLETED)
+            if fut.done():
+                return len(b)
+            raise ValueError(f'request task finished early')
```

Comment:
What is the implication of this Exception? Does this show up in user logs? Is it retried at all as a transient error?

Reply:
It's not a transient error and is not retried. It could show up anywhere that someone tries to use aiogoogle. That includes input and output of batch.

As long as the client of this code is correctly calling `close` on an InsertObjectStream, then you'll see that while handling this ValueError, you encountered the error produced by the _request_task, which is the actual cause of all this. As a result, if you're doing something like:

```
async def foo():
    async with await fs.create(...) as obj:
        await obj.write(...)
await retry_transient_errors(foo)
```
The retry_transient_errors will see the transient error from the _request_task (which will have the ValueError as a suppressed exception) and will appropriately retry foo.